### ISSUE TITLE
swaywm (xwayland): app eats 100% CPU without drawing until window changes first time

### ISSUE THREAD

**Issue by adabru** [panda3d/panda3d#1087]
## Description

I start a panda3d program via command line. The started application window pops up as expected and one first frame is rendered. Then no further frames are rendered and one python thread keeps busy at 100% CPU load.

As soon as I resize the window or change the window focus from another xwayland window to the panda3d window (doesn't work with native wayland windows), further frames are drawn as expected and the CPU isn't as busy anymore.

I tried the python debugger to narrow down the problem. The line I stopped at was https://github.com/panda3d/panda3d/blob/v1.10.8/direct/src/showbase/ShowBase.py#L2153 . I embraced the line (`self.graphicsEngine.renderFrame()`) with time measuring code and observed that before the first window change, the delta was constantly around 0.00001xxx, after the first window change the delta was constantly around  0.016xxx (=60fps) seconds.

## Steps to Reproduce

Run following code inside the sway window manager:

```python
from direct.showbase.ShowBase import ShowBase


class MyApp(ShowBase):
    def __init__(self):
        ShowBase.__init__(self)
        self.setFrameRateMeter(True)


app = MyApp()
app.run()
```

- Observe the fps meter (=stuck at its first 0.0) and your CPU load (=one CPU-core at 100%).
- Alternatively resize the window or focus another x window and then the panda3d window again. 
- Observe the fps meter (=updating around 60fps) and your CPU load (~10-20% for one CPU-core).

## Environment
 * Operating system: Arch Linux
 * System architecture: x64
 * Panda3D version: 1.10.8 (same with 1.10.7) 
 * Installation method: `pip install --user panda3d`
 * Python version (if using Python): 3.9.1
 * Compiler (if using C++): -

**Comment by rdb** [comment]
This might be because of the `_awaiting_configure` check in `x11GraphicsWindow::begin_frame()`, which early-outs until a ConfigureNotify event has been received.  Perhaps XWayland behaves unexpectedly in this respect.

**Comment by Moguri** [comment]
I can confirm that we are stuck on `_awaiting_configure` in `x11GraphicsWindow::begin_frame()`. It seems something with either Sway and/or xwayland does not always send Configure Notify events as expected (or at least the same way as X11). Doing some googling brings up some results with OBS running into issues with Sway and xwayland.

**Comment by Moguri** [comment]
Digging a bit deeper, it looks like [XReconfigureWMWindow()](https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/XReconfigureWMWindow.html) can return a non-zero value to indicate an error. In this case, I am getting `1` back. In other words, we failed to successfully send the request/event, so I do not think it makes sense to wait for a configure event in this case. However, I don't know what kind of bad things happen for Panda if it doesn't get sent. The bellow snippet used at the bottom of `x11GraphicsWindow::set_properties_now()` gets demos working for me, but I do not know if something like this would work as a legitimate fix.

```cxx
if (value_mask != 0) {                                                               
  // We must call this after changing the WM properties, otherwise we may            
  // get misleading ConfigureNotify events in the wrong order.                       
  if (XReconfigureWMWindow(_display, _xwindow, _screen, value_mask, &changes) == 0) {
    // Don't draw anything until this is done reconfiguring.                         
    _awaiting_configure = true;                                                      
  } else {                                                                           
    x11display_cat.debug()                                                           
      << "XReconfigureWMWindow() was not successful, skipping waiting on configure"; 
  }                                                                                  
}                                                                                    
```

**Comment by adabru** [comment]
Better badly fixed than not fixed at all I guess? 

@Moguri do you have a suggestion on how I can try out your fix? I didn't compile panda3d before.

**Comment by Moguri** [comment]
I'm hesitant to add any hacks like this since it may break corgi WMs. However, I misread the description, and a non-zero return value is a success, not a failure. So, I don't know why Sway/xwayland is not responding to it.

**Comment by rdb** [comment]
In fact, `XReconfigureWMWindow` always returns `True`, so its return value is not useful to us:
https://github.com/mirror/libX11/blob/5faa8dc0b364f55f19034c370436074646d9ad54/src/ReconfWM.c#L54

I was reminded of this issue when investigating a similar issue here [here](https://discourse.panda3d.org/t/fullscreen-window-in-x-loses-keyboard-focus/27801), which was due to the fact that `XReconfigureWMWindow` will not actually send a `ConfigureNotify` if nothing changed.  This could be the case if the properties happen to already be set, or if the WM decided to discard the request.

I think the `_awaiting_configure` check is therefore flawed because we can't actually be sure that a `ConfigureNotify` is sent.  We need to replace it with something better: either using Xlib's synchronization system to know that an event has been processed or instead responding to some other event like [MapNotify](https://tronche.com/gui/x/xlib/events/window-state-change/map.html) or [Expose](https://tronche.com/gui/x/xlib/events/exposure/expose.html).

**Comment by janEntikan** [comment]
I have a similar issue on X11 on FreeBSD 13.0-RELEASE since switching to WindowMaker from RatPoison as window manager. The first frame is drawn, and it doesn't get refreshed/flipped until I change the window's position or dimension. This reminded me I've also had this problem on PuppyLinux on a i363 laptop (installed using an old iso from around 2015). 

For now a quick work-around seems to be to explicitly change the window using WindowProperties:
```
base = ShowBase()
wp = WindowProperties()
wp.set_size(1280, 720)
base.win.request_properties(wp)
base.run()
```

This might not be an issue for tiling window managers because these tend to want to set their own window dimensions.

**Comment by Moguri** [comment]
@rdb, if you have something a bit more specific to try, I don't mind doing the implementation and testing. I've tried doing some digging on this topic a couple of times, but my X11 knowledge is pretty weak, and I've forgotten most of the little that I did know.

I would like to try using sway/Wayland more and this issue is the main thing holding me back. I've even been tempted to just write an sdldisplay just to work around the issue without needing to dig through X11 stuff.

**Comment by rdb** [comment]
It does send an initial configure event, but because Panda notices that the window manager has refused some of the properties, it does one more properties request to try and see if it will honour the size request.

So the fix in in #1267 would not be sufficient because the window will still freeze on all subsequent attempted size changes.

So the problem seems to be that we cannot rely on a ConfigureNotify event being sent after a configure request because the window manager may refuse all the requests.

I'm thinking maybe we need to just check that the server has had a *chance* to process the event, either by inserting some kind of fence in the event queue (I don't know whether the Xlib API has such a thing, but maybe it can be done via a custom event), or by forcibly doing an XSync.

**Comment by rdb** [comment]
I tried the "fence" method using a custom event but it doesn't actually work since the fence comes in way before the actual ConfigureNotify on my regular X11 implementation.

XSync also doesn't work.  I conjecture that it will cause the event to be delivered to the WM synchronously, but that is not a guarantee that it will be handled and returned.

There is also the glXWaitX call, which ensures that GLX calls will wait for X reconfigure requests to be done, but I think it might not be enough since we have no knowledge of what size the window actually is by the time the glViewport call gets through.  In any case, it seems to be less strong than XSync.

I checked what SDL does and they have this yucky loop in their resize-window call that waits for 100ms to see if the resize happened.  Ewww!
https://github.com/libsdl-org/SDL/blob/64c93f7804196e74e365570a01f25bfd2c0199b4/src/video/x11/SDL_x11window.c#L983-L1006

I think our fix is going to have to be similarly hacky.  I'm inclined to just have Panda time-out the _awaiting_configure state after a while, to be honest.  On my regular X11 implementation, sometimes it takes a couple of frames.  Even if I call XSync, although that does speed it up a *bit*.  I'm not going to have `set_properties_now` busy-spin like SDL though, ew.

I am inclined to also file a bug with swaywm, to be honest; I don't know if it is strictly in violation, but it is certainly very annoying behaviour that is different from other window managers.

**Comment by rdb** [comment]
Workaround implemented.  In the worst case Panda will only be consuming 100% CPU for 100 ms.  If this is considered unacceptable we will have to keep thinking about a better fix.  In any case we could consider a riskier fix for 1.11.

If you want to try it out, buildbot builds for 1.10.13 are [here](https://buildbot.panda3d.org/downloads/11746457ff31f373ff7ae25adb365f334443230f) and for master will shortly appear [here](https://buildbot.panda3d.org/downloads/63017864ab978a4f4ff69f60584677c7dcf13eb0).
